### DISCUSSION TITLE
Crash when computing a centroid

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#7005] [Q&A] [answered]
Hello,

There's a crash of the R session when I compute the centroid of the mesh in the attached file. I'm going to prepare a reproducible C++ code in order that you can try.

![crash_centroid](https://user-images.githubusercontent.com/4466543/199718677-9d0c59be-774e-424f-91dd-0abf494b727b.png)

[crash_centroid.off.txt](https://github.com/CGAL/cgal/files/9928200/crash_centroid.off.txt)

**Comment by stla** [comment]
The code:

```cpp
#include <CGAL/Exact_predicates_exact_constructions_kernel.h>
#include <CGAL/Kernel/global_functions.h>
#include <CGAL/Surface_mesh.h>
#include <CGAL/number_utils.h>
#include <CGAL/Polygon_mesh_processing/measure.h>
#include <CGAL/Polygon_mesh_processing/IO/polygon_mesh_io.h>

typedef CGAL::Exact_predicates_exact_constructions_kernel EK;
typedef EK::Point_3                                       EPoint3;
typedef CGAL::Surface_mesh<EPoint3>                       EMesh3;

namespace PMP = CGAL::Polygon_mesh_processing;

int main() {
  EMesh3 mesh;
  const bool ok = PMP::IO::read_polygon_mesh(
    "crash_centroid.off", mesh, CGAL::parameters::verbose(true)
  );
  if(!ok) {
    std::cout << "Reading failure";
    std::exit(1);
  }
  const bool valid = mesh.is_valid(false);
  if(!valid) {
    std::cout << "The mesh is not valid.";
    std::exit(1);
  }
  const EPoint3 centroid = PMP::centroid(mesh);
  std::cout << CGAL::to_double<EK::FT>(centroid.x()) << " "
            << CGAL::to_double<EK::FT>(centroid.y()) << " "
            << CGAL::to_double<EK::FT>(centroid.z()) << "\n";
  return 0;
}
```

Not sure I included all the required `.h` files (I don't know how to run this code).

**Comment by stla** [comment]
I get:

> Erreur : pile C débordée vers le haut

**Reply by lrineau** [reply]
`PMP::centroid` is probably an issue similar to the one that was fixed by @janetournois with that commit https://github.com/CGAL/cgal/pull/957/commits/5b41ca088bd2b04bed23916a7ccb15716324564e.

The issue is that the `CGAL::Epeck` kernel maintains a graph of construction, and probably `PMP::centroid` creates a long chain of constructions, by accumulating values.

Jane, do you remember that issue?

**Comment by sloriot** [comment, marked as answer]
@afabri since EPECK is used I suspect that's the same kind of issues with iterative computations (like what we fixed in area for example)

**Reply by lrineau** [reply]
We answered almost the same answer, almost at the same moment. See https://github.com/CGAL/cgal/discussions/7005#discussioncomment-4057805.

**Comment by stla** [comment]
Oddly, there's no crash with Valgrind.

If this is due to EPECK I will convert to EPICK before the calculation.

**Comment by janetournois** [comment]
Note we have already talked about similar issues https://github.com/CGAL/cgal/issues/2066 and https://github.com/CGAL/cgal/issues/2073

**Comment by stla** [comment]
Works with EPICK. Thanks.
